api/src/main/java/org/openmrs/module/openconceptlab/importer/Importer.java

```diff
@@ -113,82 +175,179 @@ public void setZipFile(ZipFile zipFile) {
+						}
 
-                if (updatedSince == null) {
-                    oclResponse = oclClient.fetchOclConcepts(subscription.getUrl(), subscription.getToken());
-					importService.updateReleaseVersion(anImport,
-                            oclClient.getOclReleaseVersion(subscription.getUrl(), subscription.getToken()));
-                } else {
-                    if (subscription.isSubscribedToSnapshot()) {
-                        oclResponse = oclClient.fetchSnapshotUpdates(subscription.getUrl(), subscription.getToken(),
-                                updatedSince);
-                    }
-                    else {
-                        oclResponse = oclClient.fetchOclConcepts(subscription.getUrl(), subscription.getToken(), lastImport.getReleaseVersion());
-						importService.updateReleaseVersion(anImport,
-                                oclClient.getOclReleaseVersion(subscription.getUrl(), subscription.getToken()));
-                    }
-                }
-
-				if (oclResponse != null) {
-					importService.updateOclDateStarted(anImport, oclResponse.getUpdatedTo());
-					importService.updateSubscriptionUrl(anImport, subscription.getUrl());
-					in = new CountingInputStream(oclResponse.getContentStream());
-					totalBytesToProcess = oclResponse.getContentLength();
-
-					processInput();
-
-					in.close();
-				}
-            }
-        });
-    }
+						// Next, read the file into a List of OclMapping and add these to the import task
+						try {
+							openInputStream();
+							JsonParser mappingsParser = objectMapper.getJsonFactory().createJsonParser(in);
+							JsonToken jsonToken = advanceToListOf("mappings", null, mappingsParser);
+							if (jsonToken != JsonToken.END_OBJECT) {
+								List<OclMapping> oclMappings = new ArrayList<>(BATCH_SIZE);
+								while (mappingsParser.nextToken() != JsonToken.END_ARRAY) {
+									OclMapping oclMapping = mappingsParser.readValueAs(OclMapping.class);
+									oclMapping.setUrl(prependBaseUrl(baseUrl, oclMapping.getUrl()));
+									oclMapping.setFromConceptUrl(prependBaseUrl(baseUrl, oclMapping.getFromConceptUrl()));
+									oclMapping.setFromSourceUrl(prependBaseUrl(baseUrl, oclMapping.getFromSourceUrl()));
+									oclMapping.setToConceptUrl(prependBaseUrl(baseUrl, oclMapping.getToConceptUrl()));
+									// We do not import Q-AND-A mappings for which the imported concept is an answer
+									boolean isQuestionAnswer = "Q-AND-A".equals(oclMapping.getMapType());
+									if (isQuestionAnswer && oclMapping.getToConceptUrl().equals(oclConcept.getUrl())) {
+										log.debug("Skipping Q-AND-A mapping for which the imported concept is a mapped answer");
+									}
+									else if (oclMapping.getToSourceName().equals(oclConcept.getSource()) && oclMapping.getToConceptCode().equals(oclConcept.getId())) {
+										log.debug("Skipping SAME-AS Mapping for OCL ID as this is added in the ImportTask, and would result in duplicate mappings");
+									}

```

Comment:
I'll be honest: this probably a guard we should have in the `ImportTask` rather than here.

Reply:
Right... the custom code was added to work around the fact that the import of CIEL wasn't including those SAME-AS maps (and to deal with the fact that dictionaries don't necessarily have the SAME-AS mapping to their concept id.